pkg/apis/discovery/types.go

```diff
@@ -114,8 +114,22 @@ type EndpointConditions struct {
 	// ready indicates that this endpoint is prepared to receive traffic,
 	// according to whatever system is managing the endpoint. A nil value
 	// indicates an unknown state. In most cases consumers should interpret this
-	// unknown state as ready.
+	// unknown state as ready. For compatibility reasons, ready should never be
+	// "true" for terminating endpoints.
 	Ready *bool
+
+	// accepting is identical to ready except it is set regardless of the terminating
+	// state of endpoints. A ready endpoint that is terminating should set this condition
+	// to true. A nil vaue indicates an unknown state. Consumers should interepret this
+	// unknown state to mean the endpoint is accepting traffic.
+	// +optional
+	Accepting *bool
+
+	// terminating indicates that this endpoint is terminating. A nil value indicates an
+	// unknown state. Consumers should interpret this unknown state to mean that the
+	// endpoint is not terminating.
```

Comment:
Probably not a good idea, but could we update the `Terminating` condition to mean both `Terminating` and `Ready`? We would simply exclude endpoints that were both terminating and not ready in this list.

Reply:
If we did that, why do we have the Ready bit at all?

Reply:
I guess I was just trying to see if we could survive without the `Serving` bool. Here's how I understand the rationale for that field:

- Adding the Terminating condition means the endpoints will stick around longer than previously.
- To make that relatively safe, we need the existing `Ready` field to be false.
- Consumers want to know if an endpoint is `Terminating` but still able to handle traffic.
- Our solution to that was to add a `Serving` field that would be true even if an endpoint was `Terminating`.

My question is if we can have a single condition that indicates that an endpoint is terminating and can handle traffic. My argument is that consumers likely don't care about endpoints that are terminating and can't handle traffic so we could just drop them from the list. (I'm not actually sure on that part, maybe consumers do care about those endpoints).

This would be a pretty simple approach:

- `Ready` remains unchanged
- New `Terminating` condition to indicate that an endpoint is terminating but still able to handle traffic. 

Maybe the actual name of the condition might need to be clearer, but if we can safely say that no one cares about the `Terminating=true,Serving=false` condition, it seems unnecessary to add a second condition here.

Reply:
> My argument is that consumers likely don't care about endpoints that are terminating and can't handle traffic so we could just drop them from the list. (I'm not actually sure on that part, maybe consumers do care about those endpoints).

While I agree that generally we won't care too much about {!R,T} endpoints, I think it would still be useful to allow consumers to check for this state. In this thread https://github.com/kubernetes/enhancements/pull/1607#discussion_r412581130 I think we agreed that while it's not very common, it's may still be useful for consumers (e.g. kube-proxy) to track this state.

Reply:
Re-reading the discussion linked, we did agree that we should explore all 4 quadrants.

@robscott if `Ready` means "serving && !terminating" and `Terminating` means "serving && terminating", then `!Ready && !Terminating` means "!serving" but that's all we know for sure. There's no representation of "!serving && terminating", right? 

It seems maybe more useful to flatten all Terminating into one.  So `Ready` means "serving && !terminating" and `Terminating` means "terminating".  It seems to me (with no real data) that pods will catch SIGTERM and either stop accepting or keep accepting, but that all pods in a set will do the same thing.  Thus, it seems unlikely that we will have a mix of serving and !serving all terminating, or at least it won't last long.  But even as I type that out, it's all async, so we WILL have that mix, but I think it will be fleeting, and it will either settle at "serving && terminating" or move from towards "!serving && terminating".  It seems unlikelty to move from "!serving" -> "serving".

For the original `Local` use-case, does this matter?  If I have an ingress LB daemonset, the difference between serving and not serving isn't going to matter much (generally 1 instance). 

So I could be convinced to ignore serving within the set of terminating.  It's not what we said before. - is the one less bit and somewhat more cognitive load worhtwhile?

Reply:
Thanks for the context here, it was helpful to read through the previous conversation around this. One of the points that stood out to me was:

> Sending traffic to any pod that fails readiness seems like we're breaking API contract. I'm personally in favor of only falling back to {R, T} endpoints

That makes it sound like there's no value in tracking endpoints that are `!Ready && Terminating`. As a variation of what Tim said above, maybe we could use 2 conditions that were defined as the following:

`ready`: `serving && !terminating`
`draining`: `serving && terminating`

Those conditions should never be true at the same time. I'm not sure that `draining` is the appropriate term here, I just wanted to throw out an alternative to `terminating` since that may be too broad for what I'm suggesting.

Reply:
Reposting Tim's comment https://github.com/kubernetes/kubernetes/pull/92968#issuecomment-720684922 cause email replies from Github doesn't seem to recognize threads:

> So the implied third state here is `!Ready && !Draining` which can only
mean !serving, regardless of terminating or not.

> This is a little different than the discussion, but I am not sure it is
meaningful.  Andrew?

Reply:
While I still think there's no real use-case yet for tracking {!R,T} endpoints, I don't think it hurts to leave room in the API to track it. If we went with `ready` and `draining` but in the future we wanted to support {!R, T}, we would have to add two additional conditions (total 4), to not breaking the existing meaning of `draining`.

I also think we want each condition to mean one thing and not have overlapping meaning with respect to terminating state. In an ideal scenario Ready refers to the pod's readiness and Terminating just refers to terminating, but because that's breaking I think the next best thing is to leave `Ready` as-is and add `serving` and `terminating`. Whether we actually account for {!R,T} in kube-proxy is a separate discussion I think.

Reply:
>  If we went with ready and draining but in the future we wanted to support {!R, T}, we would have to add two additional conditions (total 4), to not breaking the existing meaning of draining.

I think `draining` as proposed refers to a `{R, T}` state. If in the future we wanted to add a state for `{!R, T}`, I think that could be accomplished with one additional condition (`UnhealthyTerminating` or similar). That condition would also only be true if both `Ready` and `Draining` were false. I agree with you that a condition like that is probably not useful though.

It could actually be easy to explain (and understand) `ready` and `draining`, although this may be an oversimplification:

`ready`: An endpoint is ready for new connections and requests.
`draining`: An endpoint is closing out existing connections. 

On the other hand, I think a third `serving` condition may add unnecessary complexity at this point.

Reply:
> draining: An endpoint is closing out existing connections.

I think the misleading thing here is that an endpoint could be closing existing connections AND not be ready. In the context of the kube-proxy issue we're trying to address, `Draining` could mean two things:

1. an endpoint doesn't receive traffic and is closing out existing connections.
2. an endpoint is used as a fallback endpoint if no other endpoints exists.  

I think 2. will be difficult to articulate with only a `Draining` condition.

Reply:
> no real use-case yet for tracking {!R,T} endpoints

I'd amend that - there's no real use-case for tracking {!R,T} distinct from {!R,!T}.  Rob's proposal lumps those together.  If we did need to distinguish between them, we could add an "Unready" or "Failing" condition (or a better name) that explicitly means {!R, !T} (IOW - should be ready but isn't).  That leaves the "no conditions" state as implicitly {!R, T}, though I think we'd want to just call that unknown.

We're dangerously close to defining a state machine, though.

Reply:
Thinking out loud... I'll go out on a limb a bit - it can be "compatible" to split an existing set of endpoints by changing the meaning of conditions or adding new conditions as long as we believe the change is safe (won't cause things to break).

Simpy adding `terminating` is NOT safe because it first *expands* the set of endpoints to include dangerous ones (endpoints that are imminently going away) and then segments them.  As discussed elsewhere, any client who was looking just at `ready` would suddenly see more endpoints and not safe ones.

That said, once we have expanded the endpoints to "all", we can subset.  For example, if we wanted to distinguish pods that are !ready from pods that are just starting up, we could (in theory) track whether a pod has EVER been ready and, if not call it "starting", thereby subsetting the {!R} set.  Again, this sounds state-machiney, but it's just an example.  That would be a "safe" change because clients who were looking for !R would not be getting anything new.

I am trying to find other hypotheticals that would make this discussion more concrete.  Here's one.  Suppose we define a new pod-condition "flapping" - the pod is flipping ready/unready too often.

How would you think of that?

Reply:
I feel that `flapping` and `draining` are describing behaviors instead of conditions. Maybe they are the same things just at different levels of granularity.  A consumer of this API should be able to extract `flapping` by seeing the `serving` condition flip between `true` and `false` repeatedly. In the same way, `draining` should be extracted from checking `terminating` and `serving`. To me at least, EndpointSlice is the lowest denominator here (assuming it is too expensive to watch all pods) so it should try not to be prescriptive to behaviours and only describe states that are deterministic with well defined definitions.

Reply:
> Draining could mean two things:
> 1. an endpoint doesn't receive traffic and is closing out existing connections.
> 2. an endpoint is used as a fallback endpoint if no other endpoints exists.

That's a fair point, maybe draining isn't the right word here. Maybe there's another word that can communicate the `{R,T}` state?

> there's no real use-case for tracking {!R,T} distinct from {!R,!T}

Good clarification, thanks!

> That leaves the "no conditions" state as implicitly {!R, T}, though I think we'd want to just call that unknown.

Wouldn't unknown/empty map more naturally to all conditions being false? 

> Suppose we define a new pod-condition "flapping" - the pod is flipping ready/unready too often.

I'd imagine that we would set `ready` to false and optionally add a new `flapping` condition that would also only be true if every other condition for the endpoint was false.

Reply:
> I feel that flapping and draining are describing behaviors instead of conditions. 

I'm open to different names here, I just want to avoid adding two new conditions when we only actually care about one very specific combination `{R, T}`. What if we need to add some additional condition in the future that relies on pod readiness being expanded or contracted in some way? Would we need a new condition like `serving` for that expansion as well?

Reply:
Really on the fence about this one :thinking: 

While I do think a `serving` condition (or similar) is generically useful to consumers, I agree that having to add two new conditions is not ideal. My gut tells me that consumers would want a way to track pod readiness independent of anything else but I don't really have good evidence or use-cases to back this up. Since we are going to alpha gate this for one release anyways I think I'm okay to stick with one condition for now since it'll be easier to transition to two conditions if we change our minds later.

Regarding naming of the condition, I thought about both `draining` and `cordoning` but both I feel are inaccurate because they're prescriptive to the behavior consumers _should_ be doing as opposed to describing the endpoint status and letting consumers do what they wish with it. I'm thinking the best name for this condition is just `readyTerminating`. Thoughts?

Reply:
I'm not 100% sold on that name, but I don't hate it. I think `readyTerminating` makes sense if EndpointSlices were only ever going to be used to reference Pods. Since we want this to be a more generic API, I think a more generic term here might be helpful. I still don't mind `draining` but can see the hesitation there as it could be seen as too prescriptive. Naming things is hard.

Reply:
>  Since we are going to alpha gate this for one release 

At least.  If we get it wrong it could be several.

I really don't like the evolution towards N mutually-exclusive state-conditions.  I think the right answer at this layer is orthogonal.

What about "up" and "terminating" if we don't like Serving?

Reply:
> I really don't like the evolution towards N mutually-exclusive state-conditions. I think the right answer at this layer is orthogonal.

I agree.

> What about "up" and "terminating" if we don't like Serving?

I also suggested here https://github.com/kubernetes/kubernetes/pull/92968#issuecomment-720710258 to use `ready` and `readiness`. But I think @robscott's concern is less about naming (up vs serving) but more so that we are adding two conditions for one use-case.